mersenne-31/src/x86_64_avx512/poseidon2.rs

```diff
@@ -1,58 +1,303 @@
+            .iter()
+            .map(|constant| convert_to_vec_neg_form(constant.value as i32))
+            .collect();
+        Self {
+            internal_constants,
+            packed_internal_constants,
+        }
+    }
+}
+
+impl<const WIDTH: usize> Poseidon2ExternalLayerMersenne31<WIDTH> {
+    /// Construct an instance of Poseidon2ExternalLayerMersenne31 from a array of
+    /// vectors containing the constants for each round. Internally, the constants
+    ///  are transformed into the {-P, ..., 0} representation instead of the standard {0, ..., P} one.
+    fn new_from_constants(external_constants: ExternalLayerConstants<Mersenne31, WIDTH>) -> Self {
+        let packed_initial_external_constants = external_constants
+            .get_initial_constants()
+            .iter()
+            .map(|array| array.map(|constant| convert_to_vec_neg_form(constant.value as i32)))
+            .collect();
+        let packed_terminal_external_constants = external_constants
+            .get_terminal_constants()
+            .iter()
+            .map(|array| array.map(|constant| convert_to_vec_neg_form(constant.value as i32)))
+            .collect();
+        Self {
+            external_constants,
+            packed_initial_external_constants,
+            packed_terminal_external_constants,
+        }
+    }
+}
+
+/// Compute the map x -> 2^I x on Mersenne-31 field elements.
+/// x must be represented as a value in {0..P}.
+/// This requires 2 generic parameters, I and I_PRIME satisfying I + I_PRIME = 31.
+/// If the inputs do not conform to this representations, the result is undefined.
+#[inline(always)]
+fn mul_2exp_i<const I: u32, const I_PRIME: u32>(
+    val: PackedMersenne31AVX512,
+) -> PackedMersenne31AVX512 {
+    assert_eq!(I + I_PRIME, 31);
+    unsafe {
+        // Safety: If this code got compiled then AVX512-F intrinsics are available.
+        let input = val.to_vector();
+
+        // In M31, multiplication by 2^n corresponds to a cyclic rotation which
+        // is much faster than the naive multiplication method.
+
+        // Shift the low bits up. This also shifts something unwanted into
+        // the sign bit so we mark it dirty.
+        let hi_bits_dirty = x86_64::_mm512_slli_epi32::<I>(input);
+
+        // Shift the high bits down.
+        let lo_bits = x86_64::_mm512_srli_epi32::<I_PRIME>(input);
+
+        // Clear the sign bit and combine the lo and high bits.
+        // The simplest description of the operation we want is lo OR (hi_dirty AND P) which has bit pattern:
+        // 111 => 1, 110 => 1, 101 => 1, 100 => 1, 011 => 1, 010 => 0, 001 => 0, 000 => 0
+        // Note that the input patters: 111, 110, 100 cannot occur so any constant of the form **1*1000 should work.
+        let output =
+            x86_64::_mm512_ternarylogic_epi32::<0b11111000>(lo_bits, hi_bits_dirty, P_AVX512);
+        PackedMersenne31AVX512::from_vector(output)
     }
 }
```

Comment:
Hmmm, I think it should be possible to do some of these faster. Lemme think about that

Reply:
A